File: master_plan_it/master_plan_it/report/mpit_renewals_and_commitments/mpit_renewals_and_commitments.py
from __future__ import annotations

import frappe
from frappe import _

from master_plan_it.master_plan_it.financial_engine import get_renewals_commitments_dataset


def execute(filters=None):
    filters = frappe._dict(filters or {})
    dataset = get_renewals_commitments_dataset(filters)
    return _get_columns(filters), dataset.get("rows", []), None, dataset.get("chart"), _get_report_summary(dataset.get("summary", {}))


def _get_columns(filters) -> list[dict]:
    return [
        {"label": _("Contract"), "fieldname": "contract", "fieldtype": "Link", "options": "MPIT Contract", "width": 160},
        {"label": _("Title"), "fieldname": "title", "fieldtype": "Data", "width": 220},
        {"label": _("Vendor"), "fieldname": "vendor", "fieldtype": "Link", "options": "MPIT Vendor", "width": 140},
        {"label": _("Cost Center"), "fieldname": "cost_center", "fieldtype": "Link", "options": "MPIT Cost Center", "width": 140},
        {"label": _("Renewal Date"), "fieldname": "next_renewal_date", "fieldtype": "Date", "width": 120},
        {"label": _("Days"), "fieldname": "days_to_renewal", "fieldtype": "Int", "width": 80},
        {"label": _("Current Annual Amount"), "fieldname": "annual_amount_current_year", "fieldtype": "Currency", "width": 170},
        {"label": _("Next Annual Amount"), "fieldname": "annual_amount_next_year", "fieldtype": "Currency", "width": 160},
        {"label": _("Auto Renew"), "fieldname": "auto_renew", "fieldtype": "Check", "width": 100},
        {"label": _("Action"), "fieldname": "recommended_action", "fieldtype": "Data", "width": 120},
        {"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 100},
    ]


def _get_report_summary(summary: dict) -> list[dict]:
    return [
        {"label": _("Renewals"), "value": summary.get("renewal_count", 0), "datatype": "Int", "indicator": "Grey"},
        {"label": _("Current Annual"), "value": summary.get("current_annual_amount", 0), "datatype": "Currency", "indicator": "Blue"},
        {"label": _("Next Annual"), "value": summary.get("next_annual_amount", 0), "datatype": "Currency", "indicator": "Orange"},
        {"label": _("Expired"), "value": summary.get("expired_count", 0), "datatype": "Int", "indicator": "Red"},
    ]

File: master_plan_it/master_plan_it/report/mpit_renewals_and_commitments/mpit_renewals_and_commitments.js
frappe.query_reports["MPIT Renewals And Commitments"] = {
    filters: [
        { fieldname: "from_date", label: __("From Date"), fieldtype: "Date", default: frappe.datetime.get_today() },
        { fieldname: "days", label: __("Days"), fieldtype: "Int", default: 90 },
        { fieldname: "cost_center", label: __("Cost Center"), fieldtype: "Link", options: "MPIT Cost Center" },
        { fieldname: "include_children", label: __("Include Children"), fieldtype: "Check", default: 1, depends_on: "eval:doc.cost_center" },
        { fieldname: "auto_renew_only", label: __("Auto Renew Only"), fieldtype: "Check", default: 0 },
        { fieldname: "include_past", label: __("Include Past"), fieldtype: "Check", default: 0 },
        { fieldname: "min_annual_amount", label: __("Min Annual Amount"), fieldtype: "Currency", default: 0 },
        { fieldname: "action_status", label: __("Action Status"), fieldtype: "Select", options: "All\nRenew\nReview\nRenegotiate\nExpired", default: "All" },
        { fieldname: "print_profile", label: __("Print Profile"), fieldtype: "Select", options: "Executive\nStandard\nDetailed\nAudit", default: "Standard" },
        { fieldname: "print_orientation", label: __("Print Orientation"), fieldtype: "Select", options: "Auto\nPortrait\nLandscape", default: "Auto" },
        { fieldname: "print_density", label: __("Print Density"), fieldtype: "Select", options: "Normal\nCompact", default: "Normal" },
    ],

    formatter(value, row, column, data, default_formatter) {
        const formatted = default_formatter(value, row, column, data);
        if (column.fieldname === "status" && value === "Critical") {
            return `<span class="text-danger fw-semibold">${formatted}</span>`;
        }
        if (column.fieldname === "status" && value === "Warning") {
            return `<span class="text-warning fw-semibold">${formatted}</span>`;
        }
        if (column.fieldname === "recommended_action" && ["Expired", "Renegotiate"].includes(value)) {
            return `<span class="text-danger fw-semibold">${formatted}</span>`;
        }
        return formatted;
    },
};
